### ISSUE TITLE
Changing document properties when using Template Processor or loadTemplate()

### ISSUE THREAD

**Issue by studioinforma** [PHPOffice/PHPWord#908]
Does anyone have idea how to set new document properties (setCreator, setCompany, setTitle) when template is loaded and processed? I got document properties to work fine when creating new document but can't get it to work when using loadTemplate(). Thanks!

**Comment by chris-fa** [comment]
bump, is it possible to modify document properties?

**Comment by troosan** [comment]
This is currently not possible, the TemplateProcessor only allows you to change the body and the header of the document

**Comment by icy2003** [comment]
> This is currently not possible, the TemplateProcessor only allows you to change the body and the header of the document

if it is not possible, with this problem [https://github.com/PHPOffice/PHPWord/issues/991](url), how can I change docx settings?

**Comment by icy2003** [comment]
> I guess your best option is to use applyXslStyleSheet(...)
> But it would need some code changes, as currently the settings.xml is not loaded

I have a Class that extends TemplateProcessor with my own [lib](https://github.com/icy2003/icy2003_php/blob/master/iexts/phpword/TemplateProcessor.php) to change docx settings. :)
@troosan

**Comment by troosan** [comment]
@icy2003 thanks.
I integrated your changes in the develop branch

**Comment by icy2003** [comment]
@troosan thx
actually,this is because I need the function in my work so I wrote that function,will you please add more functions about settings or others because I think templateprocesser is.so powerful and it should not be only this?I am not familiar with docx, so I dont know if there has more. I think it will be helpful

**Comment by troosan** [comment]
@icy2003 indeed, I was thinking to change this method to be more generic instead of supporting only this feature. to be followed!

**Comment by icy2003** [comment]
@troosan
I will follow it.
by the way, It is too late(5:00) in China, and I am about to sleep, good night.Thx for your branch change again

**Comment by chrisribe** [comment]
> @icy2003 thanks.
> I integrated your changes in the develop branch

Can you link to where this was integrated. I am looking into setting custom document properties and templateprocessor does not seem to support it. :(

**Comment by icy2003** [comment]
> @troosan
> I will follow it.
> by the way, It is too late(5:00) in China, and I am about to sleep, good night.Thx for your branch change again

@chrisribe emmmmm……I put it in another folder, now it is https://github.com/icy2003/php/blob/master/src/iexts/PhpOffice/PhpWord/TemplateProcessor.php
and will never change again

**Comment by chrisribe** [comment]
Thanks !

But it seems to have changed a lot or is out of date compared to the current one on the develop branch.
https://github.com/PHPOffice/PHPWord/blob/develop/src/PhpWord/TemplateProcessor.php

Where you able to set the document properties via TemplateProcessor ? An how ?
Thanks

**Comment by icy2003** [comment]
@chrisribe by using regular expression, setting operate function is only setIsUpdateFields, others are some operations for generating blocks

**Comment by icy2003** [comment]
@chrisribe I provide a function for setting xml part,you can match the string for your own function, the way I wrote setIsUpdateFields was by comparing the xml I changed before and the after,maybe it's not a good way, but it runs well

**Comment by chrisribe** [comment]
If I understand correctly you load the template then replace xml after...

It is just not clear how you can proceed to load your template, set your custom property value then save.
Do you have small pseudo code example ?

I am looking to perform this on the template then save,
PhpWord setCustomProperty:
https://github.com/PHPOffice/PHPWord/blob/4fa945561f02893263e311e4c7c745df3a5dfee0/tests/PhpWord/Writer/Word2007/Part/DocumentTest.php#L52

Sorry I do not have any experience parsing docx xml values and their values relative to the docx format.

**Comment by icy2003** [comment]
@chrisribe sorry for that......I don't know about custom property until you ask me......all I know is to use xdebug and run and test and change word document and unzip and compare xml again and again

**Comment by chrisribe** [comment]
Ah ok I get it.

That allows you to figure out what is what and how the xml stucture changes based on set values in the docx. So when using your lib you never set things like "setCompany" etc ? The code is more for debug purposes ?

Was assuming you could set properties, since this issue is about "document properties".

**Comment by icy2003** [comment]
@chrisribe setIsUpdateFields

**Comment by chrisribe** [comment]
@icy2003 
https://github.com/icy2003/php/blob/a1d96b23913d34f6e516823c1a5d5b300a4f2ac2/src/iexts/PhpOffice/PhpWord/TemplateProcessor.php#L331

Only has a true/false parameter, that's why I am having a hard time decoding your code. 

_tempDocumentSettingPart == target template (init on construct).

tempDocumentMainPart == seems to be init via setMain

That's why I wanted a pseudo example on how you use your lib :)

**Comment by icy2003** [comment]
> @icy2003
> https://github.com/icy2003/php/blob/a1d96b23913d34f6e516823c1a5d5b300a4f2ac2/src/iexts/PhpOffice/PhpWord/TemplateProcessor.php#L331
> 
> Only has a true/false parameter, that's why I am having a hard time decoding your code.
> 
> _tempDocumentSettingPart == target template (init on construct).
> 
> tempDocumentMainPart == seems to be init via setMain
> 
> That's why I wanted a pseudo example on how you use your lib :)

```php
$templateProcessor = new TemplateProcessor("xxx.docx”); 
$templateProcessor->setIsUpdateFields();
```
by the way, TemplateProcessor itself is not for setting property but for replacing content……

my class it mostly for replace a table or list by using a pair of tags like:

word document:
```
${table111}
${/table111}
```
and php:

```
$templateProcessor->setTable('table111', [
1 =>['A'=>'aaa', 'B' =>'bbb']
]);
```
the array is from iWorksheet::rangeToArray of my class:
https://github.com/icy2003/php/blob/master/src/iexts/PhpOffice/PhpSpreadsheet/Worksheet/iWorksheet.php
and…… iWorksheet is an static class from PhpOffice\PhpSpreadsheet\Worksheet::rangeToArray

in other words, I need to convert excel table to word table, you can find the differents between phpoffice  and my lib at these classes, I use similar namespace so you can find them easily

**Comment by chrisribe** [comment]
@icy2003 thanks for taking the time to explain, I get what you are doing now.

**Comment by icy2003** [comment]
@chrisribe may help you with my lib

**Comment by icy2003** [comment]
> @icy2003 thanks for taking the time to explain, I get what you are doing now.

@chrisribe if you want to match yourself, see this:
https://github.com/icy2003/php/blob/master/samples/php7preg_bug.php
it may help you if you use php7.0
and, if my lib helps you in some way, give a star thanks :)

**Comment by olivermbs** [comment]
Any update on this? I am trying to change the word document title on a document generated through TemplateProcessor, is there a way of doing this?

```
$templateProcessor = new TemplateProcessor(__DIR__ . '/Template.docx');
$templateProcessor->setTitle('Title goes here');
```

**Comment by Marre-86** [comment]
I'm also interested in this

**Comment by zonky2** [comment]
> I'm also interested in this

=> same
